cpp/src/arrow/compute/kernels/sort_to_indices.cc

```diff
@@ -139,6 +151,52 @@ class CountSorter {
   }
 };
 
+template <typename ArrowType, typename Comparator>
+class CountCompareSorter {
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+  using c_type = typename ArrowType::c_type;
+
+ public:
+  explicit CountCompareSorter(Comparator compare) : compare_sorter_(compare) {}
+
+  void Sort(int64_t* indices_begin, int64_t* indices_end, const ArrayType& values) {
+    if (values.length() >= countsort_min_len_) {
+      auto min = std::numeric_limits<c_type>::max();
+      auto max = std::numeric_limits<c_type>::min();
+
+      for (int64_t i = 0; i < values.length(); ++i) {
+        min = values.IsNull(i) ? min : std::min(min, values.Value(i));
+        max = values.IsNull(i) ? max : std::max(max, values.Value(i));
+      }
+
+      // Must be unsigned compare as (max - min) may be negative for signed data.
+      if (static_cast<uint64_t>(max - min) <= countsort_max_range_) {
```

Comment:
Will an overflow here trigger UBSAN?

Reply:
I did a simple test. It will trigger ubsan if int64 overflows.

```
#include <inttypes.h>
int main(int argc, const char *argv[]) {
  int64_t min = 1LL << 63;
  int64_t max = 0;

  uint64_t v = (max - min);  // causes ubsan !
  return v;
}
```

`$ g++ -fsanitize=undefined ubsan.cc -o ubsan`
`$ ./ubsan`
`_ubsan.cc:6:21: runtime error: signed integer overflow: 0 - -9223372036854775808 cannot be represented in type 'long int'_`

Reply:
Done.
Change "(uint64)(int1 - int2)" to "(uint64)int1 - (uint64)int2". unsigned int overflow is allowed and not checked by ubsan by default.